### PR TITLE
Set /etc/hostname.

### PR DESCRIPTION
**Pull request by Random-Liu**
Fixes https://github.com/containerd/cri/issues/1041.

/cc @ibuildthecloud
Signed-off-by: Lantao Liu <lantaol@google.com>

### CODE REVIEW

**Comment by Random-Liu** [comment]
We should cherrypick this into supported branches.

File: pkg/server/sandbox_run.go [hunk 1]
```diff
@@ -456,9 +456,21 @@ func (c *criService) generateSandboxContainerSpec(id string, config *runtime.Pod
...
 
-// setupSandboxFiles sets up necessary sandbox files including /dev/shm, /etc/hosts
-// and /etc/resolv.conf.
+// setupSandboxFiles sets up necessary sandbox files including /dev/shm, /etc/hosts,
+// /etc/resolv.conf and /etc/hostname.
 func (c *criService) setupSandboxFiles(id string, config *runtime.PodSandboxConfig) error {
+	sandboxEtcHostname := c.getSandboxHostname(id)
+	if config.GetHostname() == "" {
+		// copy host's /etc/hostname
+		if err := c.os.CopyFile(etcHostname, sandboxEtcHostname, 0644); err != nil {
```
Comment (mikebrow) [inline pkg/server/sandbox_run.go]: I thought kubelet was ensuring /etc/hostname?? Shouldn't this be an error case?

File: pkg/server/sandbox_run.go [hunk 2]
```diff
@@ -456,9 +456,21 @@ func (c *criService) generateSandboxContainerSpec(id string, config *runtime.Pod
 	return g.Config, nil
 }
 
-// setupSandboxFiles sets up necessary sandbox files including /dev/shm, /etc/hosts
-// and /etc/resolv.conf.
+// setupSandboxFiles sets up necessary sandbox files including /dev/shm, /etc/hosts,
+// /etc/resolv.conf and /etc/hostname.
```
Comment (mikebrow) [inline pkg/server/sandbox_run.go]: we should probably start flagging *nix for content not applicable to other platforms..

File: pkg/server/sandbox_run.go (hunk 1 above)
Comment (Random-Liu) [reply pkg/server/sandbox_run.go]: Kubelet doesn't do that.

File: pkg/server/sandbox_run.go (hunk 2 above)
Comment (Random-Liu) [reply pkg/server/sandbox_run.go]: We'll do that when the windows support is in.

File: pkg/server/sandbox_run.go (hunk 1 above)
Comment (mikebrow) [reply pkg/server/sandbox_run.go]: sent you a txt message to chat over this for a min

Comment (mikebrow) [reply pkg/server/sandbox_run.go]: yeah I was reading the ensure hosts code in kublet.. got off track.. thx

**Comment by Random-Liu** [comment]
@mikebrow Updated the PR. Some OS doesn't have `/etc/hostname` (COS image used in GKE), so changed the PR to always `WriteFile` with correct hostname instead of copy `/etc/hostname` from the host.

**Pull request merged**